File: java/jdbc/DtUdt.java
//***************************************************************************
// (c) Copyright IBM Corp. 2007 All rights reserved.
// 
// The following sample of source code ("Sample") is owned by International 
// Business Machines Corporation or one of its subsidiaries ("IBM") and is 
// copyrighted and licensed, not sold. You may use, copy, modify, and 
// distribute the Sample in any form without payment to IBM, for the purpose of 
// assisting you in the development of your applications.
// 
// The Sample code is provided to you on an "AS IS" basis, without warranty of 
// any kind. IBM HEREBY EXPRESSLY DISCLAIMS ALL WARRANTIES, EITHER EXPRESS OR 
// IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. Some jurisdictions do 
// not allow for the exclusion or limitation of implied warranties, so the above 
// limitations or exclusions may not apply to you. IBM shall not be liable for 
// any damages you suffer as a result of using, copying, modifying or 
// distributing the Sample, even if IBM has been advised of the possibility of 
// such damages.
//***************************************************************************
//
// SOURCE FILE NAME: DtUdt.java
//
// SAMPLE: How to create, use and drop user defined distinct types
//
// SQL statements USED:
//         CREATE DISTINCT TYPE
//         CREATE TABLE
//         DROP DISTINCT TYPE
//         DROP TABLE
//         INSERT
//         COMMIT
//
// Classes used from Util.java are:
//         Db
//         JdbcException
//
// Compile: the utility file and the source file with:
//		javac Util.java
//		javac <filename>.java
//
// Run: java <filename> [<db_name>] <username> <pwd>
//	or
//     java <filename> [<db_name>] <server_name> <port_num> <username> <pwd>
//
// OUTPUT FILE: DtUdt.out (available in the online documentation)
// Output will vary depending on the JDBC driver connectivity used.
//***************************************************************************
//
// For more information on the sample programs, see the README file.
//
// For information on developing JDBC applications, see the Application
// Development Guide.
//
// For information on using SQL statements, see the SQL Reference.
//
// For the latest information on programming, compiling, and running DB2
// applications, visit the DB2 application development website at
//     http://www.software.ibm.com/data/db2/udb/ad
//**************************************************************************/

import java.lang.*;
import java.sql.*;

class DtUdt
{
  public static void main(String argv[])
  {
    try
    {
      Db db = new Db(argv);

      System.out.println();
      System.out.println("THIS SAMPLE SHOWS HOW TO CREATE, USE AND DROP\n" +
        "USER DEFINED DISTINCT TYPES.");

      // connect to the 'sample' database
      db.connect();

      create(db.con);
      use(db.con);
      drop(db.con);

      // disconnect from the 'sample' database
      db.disconnect();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e) ;
      jdbcExc.handle();
    }
  } // main

  // This function creates a few user defined distinct types
  static void create(Connection con)
  {
    try
    {
      System.out.println();
      System.out.println(
        "----------------------------------------------------------\n" +
        "USE THE SQL STATEMENTS:\n" +
        "  CREATE DISTINCT TYPE\n" +
        "  COMMIT\n" +
        "TO CREATE UDTs.");

      System.out.println();
      System.out.println(
        "  CREATE DISTINCT TYPE udt1 AS INTEGER WITH COMPARISONS");

      Statement stmt = con.createStatement();
      stmt.executeUpdate(
        "CREATE DISTINCT TYPE udt1 AS INTEGER WITH COMPARISONS");
      stmt.close();

      System.out.println(
        "  CREATE DISTINCT TYPE udt2 AS CHAR(2) WITH COMPARISONS");

      Statement stmt1 = con.createStatement();
      stmt1.executeUpdate(
        "CREATE DISTINCT TYPE udt2 AS CHAR(2) WITH COMPARISONS");
      stmt1.close();

      System.out.println(
        "  CREATE DISTINCT TYPE udt3 AS DECIMAL(7, 2) WITH COMPARISONS");

      Statement stmt2 = con.createStatement();
      stmt2.executeUpdate(
        "CREATE DISTINCT TYPE udt3 AS DECIMAL(7, 2) WITH COMPARISONS ");
      stmt2.close();

      System.out.println("  COMMIT");
      con.commit();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e, con);
      jdbcExc.handle();
    }
  } // create

  // This function uses the user defined distinct types that we created
  // at the beginning of this program.
  static void use(Connection con)
  {
    System.out.println();
    System.out.println(
      "----------------------------------------------------------\n" +
      "USE THE SQL STATEMENTS:\n" +
      "  EXECUTE IMMEDIATE\n" +
      "  COMMIT\n" +
      "TO USE UTDs.");

    // Create a table that uses the user defined distinct types
    try
    {
      System.out.println();
      System.out.println(
        "  CREATE TABLE udt_table(col1 udt1, col2 udt2, col3 udt3)");

      Statement stmt = con.createStatement();
      stmt.executeUpdate(
        "CREATE TABLE udt_table(col1 udt1, col2 udt2, col3 udt3)");
      stmt.close();

      System.out.println("  COMMIT");
      con.commit();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e, con);
      jdbcExc.handle();
    }

    // Insert data into the table with the user defined distinct types
    try
    {
      String strStmt;
      System.out.println();
      System.out.println(
        "  INSERT INTO udt_table \n" +
        "    VALUES(CAST(77 AS udt1),\n" +
        "           CAST('ab' AS udt2),\n" +
        "           CAST(111.77 AS udt3))");

      Statement stmt1 = con.createStatement();
      stmt1.executeUpdate(
        "INSERT INTO udt_table VALUES(CAST(77 AS udt1), " +
        "                             CAST('ab' AS udt2), " +
        "                             CAST(111.77 AS udt3))");
      stmt1.close();

      System.out.println("  COMMIT");
      con.commit();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e, con);
      jdbcExc.handle();
    }

    // Drop the table with the user defined distinct types
    try
    {
      System.out.println();
      System.out.println("  DROP TABLE udt_table");

      Statement stmt2 = con.createStatement();
      stmt2.executeUpdate("DROP TABLE udt_table");
      stmt2.close();

      System.out.println("  COMMIT");
      con.commit();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e, con) ;
      jdbcExc.handle();
    }
  } // use

  // This function drops all of the user defined distinct types that
  // we created at the beginning of this program
  static void drop(Connection con)
  {
    System.out.println();
    System.out.println(
      "----------------------------------------------------------\n" +
      "USE THE SQL STATEMENTS:\n" +
      "  DROP\n" +
      "  COMMIT\n" +
      "TO DROP UDTs.");

    try
    {
      System.out.println();
      System.out.println("  DROP USER DISTINCT TYPE udt1");
      Statement stmt = con.createStatement();
      stmt.executeUpdate("DROP DISTINCT TYPE udt1");
      stmt.close();

      System.out.println("  DROP USER DISTINCT TYPE udt2");
      Statement stmt1 = con.createStatement();
      stmt1.executeUpdate("DROP DISTINCT TYPE udt2");
      stmt1.close();

      System.out.println("  DROP USER DISTINCT TYPE udt3");
      Statement stmt2 = con.createStatement();
      stmt2.executeUpdate("DROP DISTINCT TYPE udt3");
      stmt2.close();

      System.out.println("  COMMIT");
      con.commit();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e, con) ;
      jdbcExc.handle();
    }
  } // drop
} // DtUdt

File: cli/dtudt.c
/****************************************************************************
** (c) Copyright IBM Corp. 2007 All rights reserved.
** 
** The following sample of source code ("Sample") is owned by International 
** Business Machines Corporation or one of its subsidiaries ("IBM") and is 
** copyrighted and licensed, not sold. You may use, copy, modify, and 
** distribute the Sample in any form without payment to IBM, for the purpose of 
** assisting you in the development of your applications.
** 
** The Sample code is provided to you on an "AS IS" basis, without warranty of 
** any kind. IBM HEREBY EXPRESSLY DISCLAIMS ALL WARRANTIES, EITHER EXPRESS OR 
** IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
** MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. Some jurisdictions do 
** not allow for the exclusion or limitation of implied warranties, so the above 
** limitations or exclusions may not apply to you. IBM shall not be liable for 
** any damages you suffer as a result of using, copying, modifying or 
** distributing the Sample, even if IBM has been advised of the possibility of 
** such damages.
*****************************************************************************
**                                                                        
** SOURCE FILE NAME: dtudt.c                                      
**                                                                        
** SAMPLE: How to create, use, and drop user-defined distinct types.
**
** CLI FUNCTIONS USED:
**         SQLAllocHandle -- Allocate Handle
**         SQLExecDirect -- Execute a Statement Directly
**         SQLFreeHandle -- Free Handle Resources
**         SQLSetConnectAttr -- Set Connection Attributes
**
** OUTPUT FILE: dtudt.out (available in the online documentation)
*****************************************************************************
**
** For more information on the sample programs, see the README file.
**
** For information on developing CLI applications, see the CLI Guide
** and Reference.
**
** For information on using SQL statements, see the SQL Reference.
**
** For the latest information on programming, building, and running DB2 
** applications, visit the DB2 application development website: 
**     http://www.software.ibm.com/data/db2/udb/ad
****************************************************************************/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sqlcli1.h>
#include "utilcli.h" /* Header file for CLI sample code */

int UDTCreate(SQLHANDLE);
int UDTUse(SQLHANDLE);
int UDTDrop(SQLHANDLE);

int main(int argc, char *argv[])
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE henv; /* environment handle */
  SQLHANDLE hdbc; /* connection handle */

  char dbAlias[SQL_MAX_DSN_LENGTH + 1];
  char user[MAX_UID_LENGTH + 1];
  char pswd[MAX_PWD_LENGTH + 1];

  /* check the command line arguments */
  rc = CmdLineArgsCheck1(argc, argv, dbAlias, user, pswd);
  if (rc != 0)
  {
    return rc;
  }

  printf("\nTHIS SAMPLE SHOWS HOW TO CREATE, USE, AND DROP UDTs.\n");

  /* initialize the CLI application by calling a helper
     utility function defined in utilcli.c */
  rc = CLIAppInit(dbAlias,
                  user,
                  pswd,
                  &henv,
                  &hdbc,
                  (SQLPOINTER)SQL_AUTOCOMMIT_ON);
  if (rc != 0)
  {
    return rc;
  }

  rc = UDTCreate(hdbc);
  rc = UDTUse(hdbc);
  rc = UDTDrop(hdbc);

  /* terminate the CLI application by calling a helper
     utility function defined in utilcli.c */
  rc = CLIAppTerm(&henv, &hdbc, dbAlias);

  return rc;
} /* main */

/* create user-defined distinct types by issuing the
   CREATE DISTINCT TYPE SQL statement */
int UDTCreate(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */
  /* SQL statements to be executed */
  SQLCHAR *stmt1 = (SQLCHAR *)
    "CREATE DISTINCT TYPE UDT1 AS INTEGER WITH COMPARISONS";
  SQLCHAR *stmt2 = (SQLCHAR *)
    "CREATE DISTINCT TYPE UDT2 AS CHAR(2) WITH COMPARISONS";
  SQLCHAR *stmt3 = (SQLCHAR *)
    "CREATE DISTINCT TYPE UDT3 AS DECIMAL(7,2) WITH COMPARISONS";

  printf("\n-----------------------------------------------------------");
  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("TO CREATE USER-DEFINED DISTINCT TYPES:\n");

  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* create UDT1 */
  printf("\n  Directly execute the statement\n");
  printf("    %s\n", stmt1);

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt1, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* create the UDT2 */
  printf("\n  Directly execute the statement\n");
  printf("    %s\n", stmt2);

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt2, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* create the UDT3 */
  printf("\n  Directly execute the statement\n");
  printf("    %s\n", stmt3);

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt3, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  return rc;
} /* UDTCreate */

/* use user-defined distinct types */
int UDTUse(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */
  /* SQL statements to be executed */
  SQLCHAR *stmt1 = (SQLCHAR *)
    "CREATE TABLE DTUDT (Col1 UDT1, Col2 UDT2, Col3 UDT3)";
  SQLCHAR *stmt2 = (SQLCHAR *)
    "INSERT INTO DTUDT "
    "  VALUES(CAST(77 AS UDT1), "
    "         CAST('ab' AS UDT2), "
    "         CAST(111.77 AS UDT3))";

  SQLCHAR *stmt3 = (SQLCHAR *)"DROP TABLE DTUDT ";

  printf("\n-----------------------------------------------------------");
  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("TO USE USER-DEFINED DISTINCT TYPES:\n");

  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* create the test table */
  printf("\n  Directly execute the statement\n");
  printf("    CREATE TABLE DTUDT (Col1 UDT1, Col2 UDT2, Col3 UDT3)\n");

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt1, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* insert values in the test table */
  printf("\n  Directly execute the statement\n");
  printf("    INSERT INTO DTUDT VALUES\n");
  printf("      VALUES(CAST(77 AS UDT1),\n");
  printf("             CAST('ab' AS UDT2),\n");
  printf("             CAST(111.77 AS UDT3))\n");

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt2, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* drop the test table */
  printf("\n  Directly execute the statement\n");
  printf("    %s\n", stmt3);

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt3, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  return rc;
} /* UDTUse */

/* drop user-defined distinct types by issuing the
   DROP DISTINCT TYPE SQL statement */
int UDTDrop(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */
  /* SQL statements to be executed */
  SQLCHAR *stmt1 = (SQLCHAR *)"DROP DISTINCT TYPE UDT1";
  SQLCHAR *stmt2 = (SQLCHAR *)"DROP DISTINCT TYPE UDT2";
  SQLCHAR *stmt3 = (SQLCHAR *)"DROP DISTINCT TYPE UDT3";

  printf("\n-----------------------------------------------------------");
  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("TO DROP USER-DEFINED DISTINCT TYPES:\n");

  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* drop UDT1 */
  printf("\n  Directly execute the statement\n");
  printf("    %s\n", stmt1);

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt1, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* drop UDT2 */
  printf("\n  Directly execute the statement\n");
  printf("    %s\n", stmt2);

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt2, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* drop UDT3 */
  printf("\n  Directly execute the statement\n");
  printf("    %s\n", stmt3);

  /* directly execute a statement */
  cliRC = SQLExecDirect(hstmt, stmt3, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  return rc;
} /* UDTDrop */

